File: backend/app/models/production.py
from enum import Enum
from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.db.base_class import Base


class ProductionStatus(str, Enum):
    DRAFT = "draft"
    PROPOSAL_SENT = "proposal_sent"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELED = "canceled"


class Production(Base):
    __tablename__ = "productions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String, nullable=False)
    organization_id: Mapped[int] = mapped_column(Integer, ForeignKey("organizations.id"), nullable=False)
    client_id: Mapped[int] = mapped_column(Integer, ForeignKey("clients.id"), nullable=True)
    status: Mapped[ProductionStatus] = mapped_column(String, default=ProductionStatus.DRAFT)
    deadline: Mapped[DateTime] = mapped_column(DateTime, nullable=True)
    priority: Mapped[str] = mapped_column(String, nullable=True)  # Priority level (high, medium, low)
    locations: Mapped[str] = mapped_column(String, nullable=True)  # JSON string with filming locations
    filming_dates: Mapped[str] = mapped_column(String, nullable=True)  # JSON string with filming dates
    created_at: Mapped[DateTime] = mapped_column(DateTime, default=func.now())
    updated_at: Mapped[DateTime] = mapped_column(DateTime, default=func.now(), onupdate=func.now())

    # Financial fields
    subtotal: Mapped[int] = mapped_column(Integer, default=0)  # Sum of all items in cents
    discount: Mapped[int] = mapped_column(Integer, default=0)  # Discount in cents
    tax_rate: Mapped[float] = mapped_column(Float, default=0.0)  # Tax rate (e.g., 10.5 for 10.5%)
    tax_amount: Mapped[int] = mapped_column(Integer, default=0)  # Calculated tax amount in cents
    total_value: Mapped[int] = mapped_column(Integer, default=0)  # Final total in cents
    total_cost: Mapped[int] = mapped_column(Integer, default=0)  # Sum of all expenses in cents
    profit: Mapped[int] = mapped_column(Integer, default=0)  # total_value - total_cost in cents

    # Payment fields
    payment_method: Mapped[str] = mapped_column(String, nullable=True)  # PIX, Crédito, Débito, Link, Crypto, Boleto
    payment_status: Mapped[str] = mapped_column(String, default="pending")  # pending, paid, overdue
    due_date: Mapped[DateTime] = mapped_column(DateTime, nullable=True)  # Payment due date

    # Notes field for additional information
    notes: Mapped[str] = mapped_column(String, nullable=True)  # Additional notes about the production

    # Relationships with eager loading
    organization = relationship("Organization", back_populates="productions")
    client = relationship("Client", back_populates="productions")
    items = relationship("ProductionItem", back_populates="production", cascade="all, delete-orphan", lazy="selectin")
    expenses = relationship("Expense", back_populates="production", cascade="all, delete-orphan", lazy="selectin")
    crew = relationship("ProductionCrew", back_populates="production", cascade="all, delete-orphan", lazy="selectin")

File: frontend/src/types/production.ts
export type ProductionStatus = 'draft' | 'proposal_sent' | 'approved' | 'in_progress' | 'completed' | 'canceled';

export interface ProductionItemResponse {
    id: number;
    production_id: number;
    name: string;
    quantity: number;
    unit_price: number;
    total_price: number;
}

export interface ProductionCrewResponse {
    id: number;
    production_id: number;
    user_id: string; // Changed from number to string
    role: string;
    fee: number;
    full_name?: string;
}

export interface ExpenseResponse {
    id: number;
    production_id: number;
    name: string;
    value: number;
    category: string;
    paid_by?: string; // Assuming this can be optional
}

export interface ClientResponse {
    id: number;
    full_name: string;
    email: string;
    cnpj?: string;
    phone?: string;
    created_at: string;
}

export interface ProductionResponse {
    id: number;
    title: string;
    organization_id: number;
    client_id?: number;
    client?: ClientResponse;
    status: ProductionStatus;
    deadline?: string;
    shooting_sessions?: Array<{ date: string | null, location: string | null }>;
    notes?: string;
    created_at: string;
    updated_at: string;
    payment_method?: string;
    payment_status: string;
    due_date?: string;
    subtotal: number;
    discount: number;
    tax_rate: number;
    tax_amount: number;
    total_value: number;
    total_cost: number;
    profit: number;
    items: ProductionItemResponse[];
    expenses: ExpenseResponse[];
    crew: ProductionCrewResponse[];
}